File: backend/src/respira_ally/core/schemas/patient.py
"""
Patient Schemas
Pydantic models for Patient API endpoints
"""
from datetime import date, datetime
from decimal import Decimal
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


# ============================================================================
# Base Schemas
# ============================================================================

class PatientBase(BaseModel):
    """Base patient information (shared fields)"""
    name: str = Field(..., min_length=2, max_length=100, description="Patient full name")
    birth_date: date = Field(..., description="Date of birth")
    gender: Literal["MALE", "FEMALE", "OTHER"] | None = Field(None, description="Gender")


# ============================================================================
# Request Schemas
# ============================================================================

class PatientCreate(PatientBase):
    """
    Patient Creation Request
    Therapists create patient profiles and assign themselves
    """
    therapist_id: UUID = Field(..., description="Assigned therapist ID")

    # Optional fields for detailed profile
    height_cm: int | None = Field(None, ge=50, le=250, description="Height in cm")
    weight_kg: Decimal | None = Field(None, ge=20, le=300, description="Weight in kg")

    # Contact information (simplified for MVP)
    phone: str | None = Field(None, max_length=20, description="Contact phone number")


class PatientUpdate(BaseModel):
    """
    Patient Update Request
    All fields optional for partial updates
    """
    name: str | None = Field(None, min_length=2, max_length=100)
    birth_date: date | None = None
    gender: Literal["MALE", "FEMALE", "OTHER"] | None = None
    height_cm: int | None = Field(None, ge=50, le=250)
    weight_kg: Decimal | None = Field(None, ge=20, le=300)
    phone: str | None = Field(None, max_length=20)
    therapist_id: UUID | None = None


class PatientQueryFilters(BaseModel):
    """
    Patient Query Filters
    Optional filters for patient list endpoint
    """
    # Search
    search: str | None = Field(None, description="Search by name or phone (case-insensitive)")

    # Gender filter
    gender: Literal["MALE", "FEMALE", "OTHER"] | None = Field(None, description="Filter by gender")

    # BMI range filter
    min_bmi: Decimal | None = Field(None, ge=0, le=100, description="Minimum BMI")
    max_bmi: Decimal | None = Field(None, ge=0, le=100, description="Maximum BMI")

    # Age range filter
    min_age: int | None = Field(None, ge=0, le=150, description="Minimum age")
    max_age: int | None = Field(None, ge=0, le=150, description="Maximum age")

    # Sorting
    sort_by: Literal["name", "birth_date", "bmi", "created_at"] | None = Field(
        "created_at", description="Sort field (default: created_at)"
    )
    sort_order: Literal["asc", "desc"] | None = Field(
        "desc", description="Sort order (default: desc)"
    )


# ============================================================================
# Response Schemas
# ============================================================================

class PatientResponse(PatientBase):
    """
    Patient API Response
    Returns essential patient information
    """
    user_id: UUID = Field(..., description="Patient user ID (primary key)")
    therapist_id: UUID | None = Field(None, description="Assigned therapist ID")

    # Physical metrics
    height_cm: int | None = None
    weight_kg: Decimal | None = None

    # Contact info
    phone: str | None = None

    # Computed field (if available)
    bmi: Decimal | None = Field(None, description="Body Mass Index (computed)")
    age: int | None = Field(None, description="Age in years (computed)")

    model_config = ConfigDict(from_attributes=True)


class PatientListResponse(BaseModel):
    """
    Patient List Response (with pagination)
    """
    items: list[PatientResponse]
    total: int = Field(..., description="Total number of patients")
    page: int = Field(..., description="Current page (0-indexed)")
    page_size: int = Field(..., description="Items per page")
    has_next: bool = Field(..., description="Has next page")


# ============================================================================
# Detailed Response (for future expansion)
# ============================================================================

class PatientDetailResponse(PatientResponse):
    """
    Detailed Patient Response
    Includes additional medical information (future use)
    """
    smoking_status: Literal["NEVER", "FORMER", "CURRENT"] | None = None
    smoking_years: int | None = None
    hospital_medical_record_number: str | None = None

    # Timestamps
    created_at: datetime | None = None
    updated_at: datetime | None = None

File: frontend/dashboard/lib/types/patient.ts
/**
 * Patient Types - TypeScript definitions for patient data
 * Based on backend schemas: respira_ally/core/schemas/patient.py
 */

// ============================================================================
// Enums
// ============================================================================

export enum Gender {
  MALE = 'MALE',
  FEMALE = 'FEMALE',
  OTHER = 'OTHER',
}

export enum RiskLevel {
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
  CRITICAL = 'critical',
}

// GOLD ABE Classification (Sprint 4)
export enum GoldGroup {
  A = 'A', // Low risk: CAT<10 AND mMRC<2
  B = 'B', // Medium risk: CAT>=10 OR mMRC>=2
  E = 'E', // High risk: CAT>=10 AND mMRC>=2
}

export enum COPDStage {
  STAGE_1 = 'stage_1',
  STAGE_2 = 'stage_2',
  STAGE_3 = 'stage_3',
  STAGE_4 = 'stage_4',
  UNKNOWN = 'unknown',
}

export enum SmokingStatus {
  NEVER = 'NEVER',
  FORMER = 'FORMER',
  CURRENT = 'CURRENT',
}

// ============================================================================
// Patient Base Types
// ============================================================================

export interface PatientBase {
  name: string // Full name
  birth_date: string // YYYY-MM-DD
  gender?: Gender
}

// ============================================================================
// Risk Assessment Types (Sprint 4 - GOLD ABE)
// ============================================================================

export interface RiskAssessmentSummary {
  gold_group: GoldGroup // GOLD ABE group (A, B, E)
  risk_level: RiskLevel // Mapped risk level (low, medium, high)
  risk_score: number // Mapped risk score (25, 50, 75)
  cat_score: number // CAT score (0-40)
  mmrc_grade: number // mMRC grade (0-4)
  exacerbation_count_12m: number // Exacerbations in last 12 months
  hospitalization_count_12m: number // Hospitalizations in last 12 months
  assessed_at: string // ISO 8601 timestamp
}

// ============================================================================
// Patient Response Types
// ============================================================================

export interface PatientResponse extends PatientBase {
  user_id: string // UUID
  therapist_id?: string // UUID

  // Physical metrics
  height_cm?: number
  weight_kg?: number

  // Contact info
  phone?: string

  // Exacerbation history (Sprint 4 - Risk Assessment)
  exacerbation_count_last_12m?: number
  hospitalization_count_last_12m?: number
  last_exacerbation_date?: string // YYYY-MM-DD

  // Computed fields
  bmi?: number
  age?: number

  // GOLD ABE Risk Assessment (Sprint 4)
  gold_group?: GoldGroup // GOLD ABE group (A, B, E)
  latest_risk_assessment?: RiskAssessmentSummary // Latest risk assessment details
}

export interface PatientListResponse {
  items: PatientResponse[]
  total: number
  page: number // 0-indexed
  page_size: number
  has_next: boolean
}

// ============================================================================
// Patient Detail Types (Future expansion)
// ============================================================================

export interface PatientDetailResponse extends PatientResponse {
  smoking_status?: SmokingStatus
  smoking_years?: number
  hospital_medical_record_number?: string
  created_at?: string // ISO 8601
  updated_at?: string // ISO 8601
}

// ============================================================================
// Query Parameters Types
// ============================================================================

export interface PatientsQuery {
  // Filtering
  risk_bucket?: RiskLevel // Filter by risk level
  adherence_rate_lte?: number // Filter by adherence rate <= value (0-100)
  last_active_gte?: string // Filter by last active date >= value (YYYY-MM-DD)

  // Sorting
  sort_by?: 'name' | 'age' | 'risk_level' | 'last_active' | 'adherence_rate'

  // Pagination
  skip?: number // Offset (default: 0)
  limit?: number // Items per page (default: 20, max: 100)
}

// ============================================================================
// Patient Create/Update Types
// ============================================================================

export interface PatientCreate extends PatientBase {
  therapist_id: string // UUID

  // Optional fields
  height_cm?: number
  weight_kg?: number
  phone?: string
}

export interface PatientUpdate {
  name?: string
  birth_date?: string // YYYY-MM-DD
  gender?: Gender
  height_cm?: number
  weight_kg?: number
  phone?: string
  therapist_id?: string // UUID
}
